lib/src/lexer.c

```diff
@@ -87,7 +89,11 @@ static void ts_lexer__get_lookahead(Lexer *self) {
     self->input.encoding == TSInputEncodingUTF8 ? ts_decode_utf8 :
     self->input.encoding == TSInputEncodingUTF16LE ? ts_decode_utf16_le : ts_decode_utf16_be;
 
-  self->lookahead_size = decode(chunk, size, &self->data.lookahead);
+  if (self->input.encoding == TSInputEncodingCustom) {
+    self->lookahead_size = self->input.decode(self->input.payload, chunk, size, &self->data.lookahead);
+  } else {
+    self->lookahead_size = decode(chunk, size, &self->data.lookahead);
+  }
```

Comment:
Can we unify this conditional with the one above, that initially defines `decode`?

Reply:
I don't think we can, since in the user-specified `decode` function we accept a payload argument that allows us to store arbitrary data (one use case being FFI), but the other decode functions don't have this argument

Reply:
Oh, interesting, I didn't notice that. What do you think about *not* having that payload argument? It looks like it isn't used in any of your example use cases in `parser_test.rs`, and it kinda makes sense that unicode decoding should probably be a pure function of the bytes.

In the context of an FFI binding, I don't think it would ever be a good idea to have a payload. We definitely don't want to be doing FFI calls for every single character.

Reply:
I see your point about not wanting to expose `extern "C"` functions to the Rust API. But since you shouldn't really need a payload at the Rust layer, could you avoid the need for that indirection by requiring that the `Encoding` be specified *statically* via a trait?

Another benefit to this is that I think it's pretty confusing for `decode` and `read` to take the same payload. This way, this custom `decode` doesn't have anything to do with any concept of a payload.

```rust
trait Encoding {
    fn decode(bytes: &[u8]) -> (i32, u32);
}

impl Parser {
    fn parse_custom_encoding<T, F, E>(
        &mut self,
        callback: &mut F,
        old_tree: Option<&Tree>
    )
    where
        E: Encoding,         // <- new type parameter for the custom encoding
        T: AsRef<[u8]>,
        F: FnMut(usine, Point) -> T,
    {
        // At compile time, create a C-compatible callback that calls the custom `decode` method.
        extern "C" fn decode_fn<E: Encoding>(data: *const u8, len: usize, char_len: *mut u32) -> i32 {
            let (c, len) = T::decode(std::slice::from_raw_parts(data, len));
            if let Some(char_len) = char_len.as_mut() {
                *char_len = len;
            }
            c
        }

        // ...

        let c_input = ffi::TSInput {
            payload: core::ptr::addr_of_mut!(payload).cast::<c_void>(),
            read: Some(read::<T, F>),
            encoding: ffi::TSInputEncodingCustom,

            // Use this custom decode callback
            decode: decode_fn::<T>(),
        };

        // ...
    }
}

```